File: lib/db/queries/budgets.ts
// Named budgets — the only budget model. A budget tracks a set of accounts +
// categories over a cycle, files into a budget group, and is either an expense
// limit or an income target (Goals are income budgets). Legacy per-category
// budgets and the goals table were removed in the budgets redesign.

import type { Exec } from '@/lib/db/repo';

export type BudgetType = 'expense' | 'income';

export interface BudgetRow {
  id: string;
  ledgerId: string;
  groupId: string | null;
  name: string;
  type: BudgetType;
  amount: number;
  /** Manual progress accumulator for one-shot income/goal budgets. */
  saved: number;
  carryForward: number;
  frequency: string;
  startDate: string;
  endDate: string | null;
  isRecurring: number;
  rollover: number;
  rolloverLimit: number | null;
  accountIds: string[];
  categoryIds: string[];
  tagIds: string[];
  warningPct: number;
}

const parseIds = (v: unknown): string[] => {
  if (v == null) return [];
  try {
    const arr = JSON.parse(String(v));
    return Array.isArray(arr) ? arr.filter((x): x is string => typeof x === 'string') : [];
  } catch {
    return [];
  }
};

function rowToBudget(r: Record<string, unknown>): BudgetRow {
  return {
    id: String(r.id),
    ledgerId: String(r.ledger_id),
    groupId: r.group_id == null ? null : String(r.group_id),
    name: r.name == null ? '' : String(r.name),
    type: String(r.type) === 'income' ? 'income' : 'expense',
    amount: Number(r.amount),
    saved: Number(r.saved ?? 0),
    carryForward: Number(r.carry_forward ?? 0),
    frequency: String(r.frequency),
    startDate: String(r.start_date),
    endDate: r.end_date == null ? null : String(r.end_date),
    isRecurring: Number(r.is_recurring ?? 1),
    rollover: Number(r.rollover ?? 0),
    rolloverLimit: r.rollover_limit == null ? null : Number(r.rollover_limit),
    accountIds: parseIds(r.account_ids),
    categoryIds: parseIds(r.category_ids),
    tagIds: parseIds(r.tag_ids),
    warningPct: Number(r.warning_pct ?? 80),
  };
}

/** List all named budgets, optionally scoped to a ledger. */
export async function listBudgets(exec: Exec, ledgerId?: string): Promise<BudgetRow[]> {
  const rows = await exec(
    ledgerId
      ? 'SELECT * FROM budgets WHERE ledger_id = ? ORDER BY created_at'
      : 'SELECT * FROM budgets ORDER BY ledger_id, created_at',
    ledgerId ? [ledgerId] : [],
  );
  return rows.map(rowToBudget);
}

export interface NewBudget {
  id: string;
  ledgerId: string;
  groupId?: string | null;
  name: string;
  type: BudgetType;
  amount: number;
  saved?: number;
  frequency: string;
  startDate: string;
  endDate?: string | null;
  isRecurring?: number;
  rollover?: number;
  rolloverLimit?: number | null;
  accountIds?: string[];
  categoryIds?: string[];
  tagIds?: string[];
  warningPct?: number;
}

const idsToJson = (ids?: string[]): string | null => (ids && ids.length ? JSON.stringify(ids) : null);

/** Insert a named budget. */
export async function createBudget(exec: Exec, b: NewBudget): Promise<void> {
  await exec(
    `INSERT INTO budgets
       (id, ledger_id, group_id, name, type, amount, saved, carry_forward,
        frequency, start_date, end_date, is_recurring, rollover, rollover_limit,
        account_ids, category_ids, tag_ids, warning_pct, created_at, updated_at)
     VALUES (?,?,?,?,?,?,?,0,?,?,?,?,?,?,?,?,?,?,datetime('now'),datetime('now'))`,
    [
      b.id,
      b.ledgerId,
      b.groupId ?? null,
      b.name,
      b.type,
      b.amount,
      b.saved ?? 0,
      b.frequency,
      b.startDate,
      b.endDate ?? null,
      b.isRecurring ?? (b.type === 'income' ? 0 : 1),
      b.rollover ?? 0,
      b.rolloverLimit ?? null,
      idsToJson(b.accountIds),
      idsToJson(b.categoryIds),
      idsToJson(b.tagIds),
      b.warningPct ?? 80,
    ],
  );
}

export interface BudgetPatch {
  groupId?: string | null;
  name?: string;
  type?: BudgetType;
  amount?: number;
  frequency?: string;
  startDate?: string;
  endDate?: string | null;
  isRecurring?: number;
  rollover?: number;
  rolloverLimit?: number | null;
  accountIds?: string[];
  categoryIds?: string[];
  tagIds?: string[];
  warningPct?: number;
}

const BUDGET_PATCH_COLUMNS: Record<keyof BudgetPatch, string> = {
  groupId: 'group_id',
  name: 'name',
  type: 'type',
  amount: 'amount',
  frequency: 'frequency',
  startDate: 'start_date',
  endDate: 'end_date',
  isRecurring: 'is_recurring',
  rollover: 'rollover',
  rolloverLimit: 'rollover_limit',
  accountIds: 'account_ids',
  categoryIds: 'category_ids',
  tagIds: 'tag_ids',
  warningPct: 'warning_pct',
};

const ARRAY_PATCH_KEYS: ReadonlySet<keyof BudgetPatch> = new Set(['accountIds', 'categoryIds', 'tagIds']);

export async function updateBudget(exec: Exec, id: string, patch: BudgetPatch): Promise<void> {
  const sets: string[] = [];
  const bind: (string | number | null)[] = [];
  for (const key of Object.keys(patch) as (keyof BudgetPatch)[]) {
    const value = patch[key];
    if (value === undefined) continue;
    sets.push(`${BUDGET_PATCH_COLUMNS[key]} = ?`);
    if (ARRAY_PATCH_KEYS.has(key)) bind.push(idsToJson(value as string[]));
    else bind.push(value as string | number | null);
  }
  if (!sets.length) return;
  sets.push("updated_at = datetime('now')");
  bind.push(id);
  await exec(`UPDATE budgets SET ${sets.join(', ')} WHERE id = ?`, bind);
}

/** Add to (or subtract from) an income/goal budget's manual `saved`, clamped at 0. */
export async function contributeBudget(exec: Exec, id: string, amount: number): Promise<void> {
  await exec(
    "UPDATE budgets SET saved = MAX(0, saved + ?), updated_at = datetime('now') WHERE id = ?",
    [amount, id],
  );
}

/** Hard-delete a named budget. */
export async function deleteBudget(exec: Exec, id: string): Promise<void> {
  await exec('DELETE FROM budgets WHERE id = ?', [id]);
}

File: ios/FinchCore/Sources/FinchCore/Store/Domain/Budgets.swift
import Foundation
import GRDB

/// Budgets domain — port of lib/db/domain/budgets/mutations.ts (the budget
/// actions; the budget-group actions live in Groups.swift). DEFERRED: the
/// withDedupMessage wrapping on create.
public enum Budgets {

    /// Reset the cached rollover (last_rolled_period → NULL, carry_forward → 0)
    /// on budgets whose category/account overlaps an edited transaction in a
    /// period at-or-before the last rolled one — so budgetProgress (which adds
    /// carry_forward) recomputes correctly. Port of lib/budgets/rollover.ts.
    /// Mainly relevant for imported web data: iOS doesn't roll budgets itself,
    /// so native budgets have a null last_rolled_period and are skipped.
    static func invalidateRollover(_ db: Database, categoryIds: [String], accountIds: [String], earliestDate: String) throws {
        guard !earliestDate.isEmpty else { return }
        for b in try Row.fetchAll(db, sql: "SELECT id, frequency, start_date, last_rolled_period, account_ids, category_ids FROM budgets WHERE last_rolled_period IS NOT NULL") {
            let lastRolled: String = b["last_rolled_period"]
            let bCats = parseIds(b["category_ids"]), bAccts = parseIds(b["account_ids"])
            let catOverlap = bCats.isEmpty || categoryIds.contains(where: bCats.contains)
            let acctOverlap = bAccts.isEmpty || accountIds.contains(where: bAccts.contains)
            guard catOverlap, acctOverlap else { continue }
            let affectedPeriod = Selectors.cycleWindow(b["frequency"], b["start_date"], earliestDate).from   // periodOf
            if affectedPeriod > lastRolled { continue }
            try db.execute(sql: "UPDATE budgets SET last_rolled_period = NULL, carry_forward = 0, updated_at = datetime('now') WHERE id = ?", arguments: [b["id"] as String])
        }
    }

    /// Compute an entry's touches (web txTouches) → feed invalidateRollover.
    /// Call BEFORE deleting an entry (it reads the entry's postings).
    static func invalidateForEntry(_ db: Database, _ entryId: String) throws {
        guard let date = try String.fetchOne(db, sql: "SELECT date FROM entries WHERE id = ?", arguments: [entryId]) else { return }
        var accountId: String?; var cats: [String] = []
        for p in try Row.fetchAll(db, sql: "SELECT account_id, category_id FROM postings WHERE entry_id = ?", arguments: [entryId]) {
            if let a = p["account_id"] as String?, accountId == nil { accountId = a }
            if let c = p["category_id"] as String? { cats.append(c) }
        }
        guard let acct = accountId else { return }
        try invalidateRollover(db, categoryIds: cats, accountIds: [acct], earliestDate: date)
    }

    private static func parseIds(_ raw: String?) -> [String] {
        guard let raw, let d = raw.data(using: .utf8), let arr = try? JSONDecoder().decode([String].self, from: d) else { return [] }
        return arr
    }

    public static let handlers: [ActionName: Apply.Handler] = [
        .createBudget: create,
        .updateBudget: update,
        .updateBudgetCycle: updateCycle,
        .clearPendingAmount: clearPending,
        .removeBudget: remove,
        .contributeBudget: contribute,
    ]

    /// JSON array text, or nil for an empty list — matches the web `idsToJson`
    /// (`ids.length ? JSON.stringify(ids) : null`), so the column stores NULL.
    private static func idsToJson(_ ids: [String]) -> String? {
        if ids.isEmpty { return nil }
        return (try? JSONEncoder().encode(ids)).flatMap { String(data: $0, encoding: .utf8) }
    }

    static func create(_ db: Database, _ args: Args) throws {
        struct A: Decodable {
            let id: String?; let ledgerId: String?; let groupId: String?; let name: String; let type: String?
            let amount: Double; let saved: Double?; let frequency: String?; let startDate: String?; let endDate: String?
            let isRecurring: Double?; let rolloverLimit: Double?
            let accountIds: [String]?; let categoryIds: [String]?; let warningPct: Double?
        }
        let a = try args.to(A.self)
        let name = a.name.trimmingCharacters(in: .whitespacesAndNewlines)
        if name.isEmpty { throw I18nError("error.required.budgetName", [:], "Budget name is required") }
        let type = a.type == "income" ? "income" : "expense"
        if !(a.amount > 0) { throw I18nError("error.budget.amountGt0", [:], "Budget amount must be greater than 0") }
        let rollover = (args.values["rollover"]?.isTruthy ?? false) ? 1 : 0
        let isRecurring = a.isRecurring.map { Int($0) } ?? (type == "income" ? 0 : 1)
        let startDate = a.startDate ?? String(ISO8601DateFormatter().string(from: Date()).prefix(10))
        // UNIQUE(ledger_id, name, frequency, start_date) → friendly dup error (web parity).
        try Dedup.wrap {
            try db.execute(sql: """
                INSERT INTO budgets (id, ledger_id, group_id, name, kind, amount, saved, carry_forward,
                    frequency, start_date, end_date, is_recurring, rollover, rollover_limit,
                    account_ids, category_ids, warning_pct, created_at, updated_at)
                VALUES (?,?,?,?,?,?,?,0,?,?,?,?,?,?,?,?,?,datetime('now'),datetime('now'))
                """, arguments: [a.id ?? Entries.newId("bgt"), a.ledgerId ?? "personal", a.groupId, name, type,
                                 a.amount, a.saved ?? 0, a.frequency ?? "monthly", startDate, a.endDate, isRecurring,
                                 rollover, a.rolloverLimit, idsToJson(a.accountIds ?? []), idsToJson(a.categoryIds ?? []),
                                 a.warningPct ?? 80])
        }
    }

    private static let cols: [String: String] = [
        "groupId": "group_id", "name": "name", "type": "kind", "amount": "amount", "frequency": "frequency",
        "startDate": "start_date", "endDate": "end_date", "isRecurring": "is_recurring", "rollover": "rollover",
        "rolloverLimit": "rollover_limit", "accountIds": "account_ids", "categoryIds": "category_ids", "warningPct": "warning_pct",
    ]

    static func update(_ db: Database, _ args: Args) throws {
        guard let id = args.idString else { throw I18nError("error.invalidArgs", [:], "updateBudget requires an id") }
        let patch = args.patchObject
        if let n = patch["name"], !n.isNonEmptyTrimmedString { throw I18nError("error.required.budgetName", [:], "Budget name is required") }
        if let amt = patch["amount"], let a = amt.asDouble, !(a > 0) { throw I18nError("error.budget.amountGt0", [:], "Budget amount must be greater than 0") }

        // Amount-only edit on a recurring budget → stage it for the next cycle.
        if patch.count == 1, let amt = patch["amount"]?.asDouble {
            if try Int.fetchOne(db, sql: "SELECT is_recurring FROM budgets WHERE id = ?", arguments: [id]) == 1 {
                try db.execute(sql: "UPDATE budgets SET pending_amount = ?, updated_at = datetime('now') WHERE id = ?", arguments: [amt, id])
                return
            }
        }

        var sets: [String] = []
        var bind: [DatabaseValueConvertible?] = []
        for (key, col) in cols where patch.keys.contains(key) {
            sets.append("\(col) = ?")
            if key == "accountIds" || key == "categoryIds" { bind.append(idsToJson(patch[key]!.asStringArray)) }
            else { bind.append(patch[key]!.sqlBind) }
        }
        if sets.isEmpty { return }
        sets.append("updated_at = datetime('now')")
        bind.append(id)
        try db.execute(sql: "UPDATE budgets SET \(sets.joined(separator: ", ")) WHERE id = ?", arguments: StatementArguments(bind))
    }

    static func updateCycle(_ db: Database, _ args: Args) throws {
        guard let id = args.idString else { throw I18nError("error.invalidArgs", [:], "updateBudgetCycle requires an id") }
        let patch = args.patchObject
        let validFreqs: Set<String> = ["daily", "weekly", "biweekly", "monthly", "quarterly", "yearly"]
        guard let frequency = patch["frequency"]?.asString, validFreqs.contains(frequency) else {
            throw I18nError("error.budget.unknownFreq", ["freq": patch["frequency"]?.asString ?? "nil"], "Unknown frequency")
        }
        guard let startDate = patch["startDate"]?.asString, isYMD(startDate) else {
            throw I18nError("error.budget.dateFormat", [:], "startDate must be YYYY-MM-DD")
        }
        if let amt = patch["amount"]?.asDouble, !(amt > 0) { throw I18nError("error.budget.amountGt0", [:], "Budget amount must be greater than 0") }
        guard let existing = try Row.fetchOne(db, sql: "SELECT amount, end_date FROM budgets WHERE id = ?", arguments: [id]) else {
            throw I18nError("error.notFound.budget", [:], "Budget not found")
        }
        let amount = patch["amount"]?.asDouble ?? (existing["amount"] as Double)
        // undefined keeps existing; explicit (incl. null) sets it.
        let endDate: String? = patch.keys.contains("endDate") ? patch["endDate"]?.asString : (existing["end_date"] as String?)
        try db.execute(sql: """
            UPDATE budgets SET amount = ?, frequency = ?, start_date = ?, end_date = ?,
                pending_amount = NULL, last_rolled_period = NULL, updated_at = datetime('now')
             WHERE id = ?
            """, arguments: [amount, frequency, startDate, endDate, id])
    }

    static func clearPending(_ db: Database, _ args: Args) throws {
        struct A: Decodable { let id: String }
        try db.execute(sql: "UPDATE budgets SET pending_amount = NULL, updated_at = datetime('now') WHERE id = ?", arguments: [try args.to(A.self).id])
    }

    static func remove(_ db: Database, _ args: Args) throws {
        struct A: Decodable { let id: String }
        try db.execute(sql: "DELETE FROM budgets WHERE id = ?", arguments: [try args.to(A.self).id])
    }

    static func contribute(_ db: Database, _ args: Args) throws {
        struct A: Decodable { let id: String; let amount: Double }
        let a = try args.to(A.self)
        if !a.amount.isFinite { throw I18nError("error.budget.invalidContribution", [:], "Invalid contribution amount") }
        try db.execute(sql: "UPDATE budgets SET saved = MAX(0, saved + ?), updated_at = datetime('now') WHERE id = ?", arguments: [a.amount, a.id])
    }
}
